components/shared/onboarding/Onboarding.tsx

```diff
@@ -0,0 +1,320 @@
+  };
+  const handleDaySelectTo = (date: Date | undefined) => {
+    if (!date) {
+      setSelectedTo(date as any);
+      return;
+    }
+
+    const newDate = new Date(
+      date.getFullYear(),
+      date.getMonth(),
+      date.getDate()
+    );
+    setSelectedTo(newDate as any);
+  };
+  const onSubmit = async () => {
+    console.log(getValues());
+    try {
+      const {
+        portfolio,
+        availability,
+        learningGoals,
+        email,
+        experiences,
+        newProjects,
+        fullname,
+      } = getValues();
+      const dbExperiences = experiences?.map((experience) => experience.name);
+      const startTime = availability?.startTime;
+      const endTime = availability?.endTime;
+      if (currentStep === "1") {
+        await updateUser({
+          updateData: {
+            portfolio,
+            email,
+            fullname,
+          },
+        });
+      } else if (currentStep === "2") {
+        await updateUser({
+          updateData: {
+            learningGoals,
+          },
+        });
+        changeStep();
+      } else if (currentStep === "3") {
+        await updateUser({
+          updateData: {
+            experiences: dbExperiences,
+          },
+        });
+        changeStep();
+      } else if (currentStep === "4") {
+        await updateUser({
+          updateData: {
+            newProjects,
+            startTime,
+            endTime,
+          },
+        });
+        router.push("/");
+      }
+    } catch (e) {
+      console.log(e);
+    }
+  };
+
+  const renderForm = () => {
+    switch (currentStep) {
+      case "1":
+        return (
+          <OnboardingOne
+            user={user}
+            step={step}
+            register={register}
+            errors={errors}
+          />
+        );
+      case "2":
+        return (
+          <OnboardingTwo
+            step={step}
+            goalFields={goalFields}
+            register={register}
+            appendGoal={appendGoal}
+            removeGoal={removeGoal}
+          />
+        );
+      case "3":
+        return (
+          <OnboardingThree
+            register={register}
+            step={step}
+            experiences={experienceFields}
+            errors={errors}
+            appendExperience={appendExperience}
+            removeExperience={removeExperience}
+          />
+        );
+      case "4":
+        return (
+          <OnboardingFour
+            register={register}
+            control={control}
+            selectedFrom={selectedFrom!}
+            selectedTo={selectedTo!}
+            handleDaySelectFrom={handleDaySelectFrom}
+            handleDaySelectTo={handleDaySelectTo}
+            errors={errors}
+            step={step}
+          />
+        );
+      default:
+        return (
+          <OnboardingOne
+            user={user}
+            step={step}
+            register={register}
+            errors={errors}
+          />
+        );
+    }
+  };
```

Comment:
I don't think you should have a renderForm. I think you should avoid using a switch when we can group logic together. Instead make an object with the components and fields. Then we can simply key into which component we need based on the step.

```js

  const stepMap: {
    [key: string]: {
      component: React.ReactNode;
      fields: string[];
    };
  } = {
    1: {
      fields: ["fullname", "portfolio", "email"],
      component: (
        <OnboardingOne
          user={user}
          step={step}
          register={register}
          errors={errors}
        />
      ),
    },
    2: {
      fields: ["learningGoals"],
      component: (
        <OnboardingTwo
          step={step}
          goalFields={goalFields}
          register={register}
          appendGoal={appendGoal}
          removeGoal={removeGoal}
        />
      ),
    },
    3: {
      fields: ["experiences"],
      component: (
        <OnboardingThree
          register={register}
          step={step}
          experiences={experienceFields}
          errors={errors}
          appendExperience={appendExperience}
          removeExperience={removeExperience}
        />
      ),
    },
    4: {
      fields: ["newProjects", "availability"],
      component: (
        <OnboardingFour
          register={register}
          control={control}
          selectedFrom={selectedFrom!}
          selectedTo={selectedTo!}
          handleDaySelectFrom={handleDaySelectFrom}
          handleDaySelectTo={handleDaySelectTo}
          errors={errors}
          step={step}
        />
      ),
    },
  };
```

Now rendering our form isn't calling a function- it's just rendering the right component:

```js
  <form
        className="flex-0 w-full overflow-y-hidden"
        onSubmit={handleSubmit(onSubmit)}
      >
        {stepMap[currentStep].component}

```

The fields part of the object we can use for our 'nextStep' function to heavily simplify things and avoid using switches.